ONYX-10152 : Design Token Schema

tzheleznyak opened the pull request:
### What does this PR do?
Design Of Token Schema
https://github.com/cyberark/conjur/blob/bdcdd99a454ef93f2858b209c96e74dcb7a98499/design/authenticators/authn_jwt/token_schema.md

### What ticket does this PR close?
ONYX-10152

### Checklists

#### Change log
- [ ] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation

#### API Changes
- [ ] The [OpenAPI spec](https://github.com/cyberark/conjur-openapi-spec) has been updated to meet new API changes (or an issue has been opened), or
- [ ] The changes in this PR do not affect the Conjur API

orenbm commented:
@tzheleznyak can you add a link to the page so it's easier to read it? @nessiLahav did it here: https://github.com/cyberark/conjur/pull/2308

eladkug reviewed (commented):
Thanks @tzheleznyak . Few questions : 
1. If the user will use mandatory_claims and will not declare claim_mapping it will be o.k.? if not how can we enforce it?
2. I think we need to bold that it will support only jwt (require specific documentation on jwt only), and verify its not affected other authenticators 
3. Do you see any security concern as part of this insert ? Maybe we need to pay attention here to our logs .

oburstein-hub commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+
+| Name             | Value                   | Description                                      |
+| ---------------- | ----------------------- | ------------------------------------------------ |
+| mandatory_claims | branch, job, aud        | List of claims as they appear in host annotation |
```
Are "branch, job, aud" only an example ? Do you have a full list of values for mandatory claims ?

tzheleznyak replied:
These are examples

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+- [Definition of Done](#definition-of-done)
+- [Solution Review](#solution-review)
+
+<small><i><a href='http://ecotrust-canada.github.io/markdown-toc/'>Table of contents generated with markdown-toc</a></i></small>
```
you can remove this :)

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+- [Table of Contents](#table-of-contents)
+- [Glossary](#glossary)
+- [Useful Links](#useful-links)
+- [Requirments](#requirments)
```
```suggestion
- [Requirments](#requirements)
```

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+  * [Flow Explanation](#flow-explanation)
+- [Backwards Compatibility](#backwards-compatibility)
+- [Affected Components](#affected-components)
+- [Delivery Plan](#delivery-plan)
+- [Delivery Plan](#delivery-plan-1)
```
two delivery plans? I would differentiate them somehow. Maybe according to different proposed flows?

tzheleznyak replied:
Good catch its mistake

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+
+| **Term**         | **Description**                                              |
+| ---------------- | ------------------------------------------------------------ |
+| JWT Claims       | claims are pieces of information asserted about a subject    |
```
```suggestion
| JWT Claims       | Claims are pieces of information asserted about a subject    |
```

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+| **Term**         | **Description**                                              |
+| ---------------- | ------------------------------------------------------------ |
+| JWT Claims       | claims are pieces of information asserted about a subject    |
+| Host Annotation  | Defining of check that we should do on authentication request for host. In JWT authentication its a claim check. |
```
```suggestion
| Host Annotation  | Defines the checks that we should do on authentication request for host. In JWT authentication its a claim check. |
```

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+| ---------------- | ------------------------------------------------------------ |
+| JWT Claims       | claims are pieces of information asserted about a subject    |
+| Host Annotation  | Defining of check that we should do on authentication request for host. In JWT authentication its a claim check. |
+| Mandatory Claims | JWT Claims that host must validate their existance and value |
```
```suggestion
| Mandatory Claims | JWT-specific claims that a host must validate their existence and value |
```

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+[//]: # "Elaborate on the issue you are writing a solution for"
+
+1. Required fields
+   1. The user will have the ability to declare in the authenticator policy which claims must be in host annotations. And therefore also in the JWT token.
```
```suggestion
   1.  In the authenticator policy, the user will have the ability to declare which claims must be in host annotations. And therefore also in the JWT token.
```

sigalsax replied:
>  And therefore also in the JWT token.

Not sure what this means. Please explain :)

tzheleznyak replied:
Any annotation in the host that authenticating again JWT authenticator we check it is a claim inside the token payload. If it doesn't we fail the authentication request. I meant here that if we verify that a claim in the host annotations it automatically means that we verify that it is in the decoded token

sigalsax replied:
I suggest we make that clear. That the claims are checked inside the token's payload

> if we verify that a claim in the host annotations it automatically means that we verify that it is in the decoded token

great! suggest we add that :)

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+
+1. Required fields
+   1. The user will have the ability to declare in the authenticator policy which claims must be in host annotations. And therefore also in the JWT token.
+   2. Upon authentication request the authenticator MUST validate those claims - existence and content 
```
```suggestion
   2. When sending an authentication request, the authenticator MUST validate those claims - existence and content 
```

sigalsax replied:
as in we check the existence of the claims and the content?

tzheleznyak replied:
Yes

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+   1. The user will have the ability to declare in the authenticator policy which claims must be in host annotations. And therefore also in the JWT token.
+   2. Upon authentication request the authenticator MUST validate those claims - existence and content 
+2. Mapping 
+   1. The user should have the ability to map a given "technical" claim name to a more familiar / business one - i.e, in git-lab - instead of mention in the policy "sub" claim, he will have the ability to mention "job"
```
```suggestion
   1. The user should have the ability to map a given "technical" claim name to a more familiar / business one - i.e, in git-lab. Instead of mentioning in the policy "sub" claim, the user will have the ability to mention "job"
```

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+## Solution
+[//]: # "Elaborate on the solution you are suggesting in this page. Address the functional requirements and the non functional requirements that this solution is addressing. If there are a few options considered for the solution, mention them and explain why the actual solution was chosen over them. Add an execution plan when relevant. It doesn't have to be a full breakdown of the feature, but ju ...
+
+User with define the token schema variables in authenticator policy. They will be checked in the JWT authentication request.
```
```suggestion
User will define the token schema variables in the authenticator policy. The variables will be checked in the JWT authentication request.
```

tzheleznyak commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+| Host Annotation  | Defining of check that we should do on authentication request for host. In JWT authentication its a claim check. |
+| Mandatory Claims | JWT Claims that host must validate their existance and value |
+| Claims Mapping   | Mapping between claim name and name of host annotation       |
+| Token Schema     | Combination of mandatory claims and claims mapping           |
```
Removed the token schema line. We can use one of them.

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+| mandatory_claims | branch, job, aud        | List of claims as they appear in host annotation |
+| claims_mapping   | {ref: branch, sub: job} | Mapping between host annotation to JWT claim     |
+
+For any claim being mapped we will add the host annotation to the `mandatory_claims` variable and not the JWT claim name.
```
```suggestion
For any claim mapped, we will add the host annotation to the `mandatory_claims` variable and not the JWT claim name.
```

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+## Design
+[//]: # "Add any diagrams, charts and explanations about the design aspect of the solution. Elaborate also about the expected user experience for the feature"
+
+The following class diagram, represent the flow of checking the mandatory claims and the claim mapping.
```
```suggestion
The following class diagram represents the flow of checking the mandatory claims and the claim mapping.
```

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+
+### Flow Explanation
+
+Upon validate_restrictions is called in JWTVendorConfiguration class these what would happen:
```
```suggestion
When validate_restrictions is called in JWTVendorConfiguration class the following will happen as a result:
```

tzheleznyak replied:
Done

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+### Class Diagrams
+[//]: # "Describe classes that are going to be added /changes and their immediate environment. Non-changed classes may be colored differently"
+
+![image-20210711122753500](token-schema-class-diagram.png)
```
personal preference,
Make the diagram's boxes closer together and the lettering bigger, otherwise even if we expand it, it will be hard to read. See Oren's or my designs for example :)

tzheleznyak replied:
Great point. Fixed it

tzheleznyak commented:
> Thanks @tzheleznyak . Few questions :
> 
> 1. If the user will use mandatory_claims and will not declare claim_mapping it will be o.k.? if not how can we enforce it?
> 2. I think we need to bold that it will support only jwt (require specific documentation on jwt only), and verify its not affected other authenticators
> 3. Do you see any security concern as part of this insert ? Maybe we need to pay attention here to our logs .

1. They are optional. Added line about it.
2. Added line about it.
3. I don't see any security concern here

benoaviram commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,178 @@
+
+Two new variables will be added to the authenticator policy:
+
+* mandatory_claims
```
We didn't talked about naming it mandatory_host_claims? 
in case we will need to have "mandatory_clims" for the authenticator level in the future.. here it is for the host level..

tzheleznyak replied:
@hilagross what are the decided names?

nessiLahav replied:
i also prefer the name of the variable will contains `host` or `app` in continue to @benoaviram comment

hilagross replied:
- host_mandatory_claims
- mapping_claims

benoaviram commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,178 @@
+| mandatory_claims | branch, job, aud        | List of claims as they appear in host annotation |
+| claims_mapping   | {ref: branch, sub: job} | Mapping between host annotation to JWT claim     |
+
+* For any claim mapped, we will add the host annotation to the `mandatory_claims` variable and not the JWT claim name.
```
I think that "use" will be better here.. we will not add it.. the user adds it.. we will use the mapped name instead

tzheleznyak replied:
?

benoaviram commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,178 @@
+
+1. The `valitate_restrictions` function will call `CreateConstraintsFromPolicy` that create constraints object checking the hosts annotations.
+   1. It will create instance of `NonEmptyConstraint` to check we have at least one host annotation.
+   2. It will call `LoadMAndatoryClaims` command class to load the list of mandatory claims for the `mandatory_claims` variable
```
a <- A

tzheleznyak replied:
?

benoaviram commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,178 @@
+   1. It will create instance of `NonEmptyConstraint` to check we have at least one host annotation.
+   2. It will call `LoadMAndatoryClaims` command class to load the list of mandatory claims for the `mandatory_claims` variable
+   3. It will create `RequiredContraint` object from it
+2. The `validate_restrictions` functon will call `CreateValidatorFromPolicy` to create validator with the claims mapping loaded
```
function*

nessiLahav replied:
Why do we need this class `CreateValidatorFromPolicy` what her responsibility ?

You could directly create the `ValidateRestrictionOneToOne` which has a dependency of key mapping.
And maybe in the future when we will extend this functionality to more possibilities for example  `ValidateRestrictionWithConditions` or `ValidateRestrictionByRegex`
So you will need this extra layer of `CreateValidatorFromPolicy` to decide which object to create 

WDYT?

benoaviram commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,178 @@
+## Backwards Compatibility
+[//]: # "How will the design of this solution impact backwards compatibility? Address how you are going to handle backwards compatibility, if necessary"
+
+Token schema is not mandatory therfore the JWT authentication will remain backwards compatible.
```
What will happen for existing hosts ? 
if the user setup the JWT authenticator, then creates Host A and B. 
then, adds the above variables, what will happen to Host A and B? 
If they will fail to authenticate... then we aren't keeping the backward compatibility ..

tzheleznyak replied:
Added this to Open Questions

sashaCher commented:
Hey,
I think that we should have a separate discussion regarding [registered claim names](https://datatracker.ietf.org/doc/html/rfc7519#page-9).
They can not be mapped for the whole verification flow and part of them are meaning less as mandatory.
Registered claim names are: `iss`, `sub`, `aud`, `exp`, `nbf`, `iat`, `jti`.

`exp`, `nbf`, `iat` and `jti` cannot be mandatory in terms of annotations: first three depends time and the last should be unique for each token. It's meaningless to require from host owner to provide the annotation related to the claims and as a result it's meaningless to map them.
From other side there's a valid use case when a customer may require to make `nbf` claim mandatory that will mean that each token has to have a valid (from time perspective point of view) `nbf` in order to be valid.

`iss` issuer is the same for all tokens of the same jwks/oidc provider uri. It's validated in 3rd party before we're able to apply mapping. `iss` should behave that same way as time related claim means only to require that it will be presented in the token equal to provider's host name or issuer variable (already has this logic).

`aud` actually can be an array so it's mandatory behaviour is `includes` rather `equals`.

InbalZilberman commented:
@tzheleznyak what is the meaning of mandatory audience? will each host will have an aud annotation?
Do we have cases of claims that are will not have a corresponding annotation?
What will be the UX of annotation that is mandatory and is not in the host's annotation? How does it reflect in your design?
What will be the UX of annotation that has mapping but it's mapped value is not found? How does it reflect in your design?

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+
+Upon validate_restrictions is called in JWTVendorConfiguration class these what would happen:
+
+1. The `valitate_restrictions` function will call `CreateConstraintsFromPolicy` that create constraints object checking the hosts annotations.
```
```suggestion
1. The `validate_restrictions` function will call `CreateConstraintsFromPolicy` that create constraints object checking the hosts annotations.
```

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+Upon validate_restrictions is called in JWTVendorConfiguration class these what would happen:
+
+1. The `valitate_restrictions` function will call `CreateConstraintsFromPolicy` that create constraints object checking the hosts annotations.
+   1. It will create instance of `NonEmptyConstraint` to check we have at least one host annotation.
```
you can remove all the `It`s and just leave `Will`

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+1. The `valitate_restrictions` function will call `CreateConstraintsFromPolicy` that create constraints object checking the hosts annotations.
+   1. It will create instance of `NonEmptyConstraint` to check we have at least one host annotation.
+   2. It will call `LoadMAndatoryClaims` command class to load the list of mandatory claims for the `mandatory_claims` variable
+   3. It will create `RequiredContraint` object from it
```
```suggestion
   3. It will create `RequiredConstraint` object from it
```

sigalsax replied:
You can just join the two sentences
>    2. Will call `LoadMAndatoryClaims` command class to load the list of mandatory claims for the `mandatory_claims` variable. This class will create and return `RequiredConstraint` object.

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+2. The `validate_restrictions` functon will call `CreateValidatorFromPolicy` to create validator with the claims mapping loaded
+   1. It will call the `LoadClaimsMapping` command class to get dictionary between claim to annotation.
+   2. It will create *ValidateRestrictionOneToOne* object and will return it
+3. `ValidateResourceRestrictions` will be called with the contraints and the validator from the previous steps. This run will check the claims in the token and annotations. If anything wrong like a missing claim in the host annotation or claim with invalid value and error with be thrown and the authentication request will be denied.
```
```suggestion
3. `ValidateResourceRestrictions` will be called with the constraints and the validator from the previous steps. This run will check the claims in the token and annotations. If anything wrong like a missing claim in the host annotation or a claim has an invalid value, an error with be thrown and the authentication request will be denied.
```

at this step do we check the integrity of the claim or just run some validations that they exist?

sigalsax replied:
what would be an example of an invalid value?

tzheleznyak replied:
for example annotation of host is 
authn-jwt/gitlab/ref: master
and ref claim is something else like "develop" so we should fail authentication

sigalsax replied:
an example would help then

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+## Backwards Compatibility
+[//]: # "How will the design of this solution impact backwards compatibility? Address how you are going to handle backwards compatibility, if necessary"
+
+Token schema is not mandatory therfore the JWT authentication will remain backwards compatible.
```
```suggestion
Token schema is not mandatory therefore the JWT authentication will remain backwards compatible.
```

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+| Support mandatory claims                         | 5 SP       |
+| Support claims mapping                           | 5 SP       |
+| Add status checks for token schema configuration | 2 SP       |
+| Add Cucumber tests for token schema              | 5 SP       |
```
what about a security review by Sec Champion and the team? _Especially_ since we are dealing with authentication

sigalsax replied:
I would highly advice that the code review is conducting live with the group to evaluate the areas of vulnerability and the sec boundary

nessiLahav replied:
What about the following sections:
Demo , Docs, Security

tzheleznyak replied:
There is nothing to write there so why to add empty sections?

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+* How the `claims_mapping` should look like
+  * Option 1: A dictionary
+    `{ref: branch, sub: job}`
+  * Option 2: Just the paris
```
paris?

`pairs`?

sigalsax commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,180 @@
+- Test plan is reviewed
+- Acceptance criteria have been met
+- Tests are implemented according to test plan 
+- The behaviour is documented in Conjur Open Source and Enterprise
```
not only that, we need to check that the flow works according to docs

sigalsax commented:
@tzheleznyak I think we should add a section about security. Some helpful questions to ask ourselves when evaluating changes we want to add/to change:
1. From a security standpoint, what is interesting in the addition or in change you plan to implement. In other words, what would those that view your doc from OSS community (and internal) want to verify is addressed before using the product.

2. What are trust boundaries?
Trust boundary describes a boundary where within it there is trust. It consists of a location in a data flow diagram where data changes its level of trust. Any place where data is passed between two different processes is typically a trust boundary. For example, a k8s namespace. The namespace is trust boundary for the secrets he reads in his namespace. The namespace in this example, is a trusted entity and all other entities beyond it are untrusted.

3. What are controls?
Things you add for mitigation of concerns. For example, problem: someone will use service account maliciously solution: service account will be created only for secrets provider validation on input.

oburstein-hub commented:
Hi Tamir
I have looked into the class diagram of the design document and I have some comments
1. Class diagram should not require zoom-in to see its details
2. Not clear what the arrows are - not according to UML
3. Should not have writing on arrows of what the action does

oburstein-hub commented:
Sequence diagram should be added to describe flow
Also - there is a template for design in Confluence. Please, verify you follow its sections

tzheleznyak commented:
> Sequence diagram should be added to describe flow
> Also - there is a template for design in Confluence. Please, verify you follow its sections

I used the design template from this project

tzheleznyak commented:
> @tzheleznyak what is the meaning of mandatory audience? will each host will have an aud annotation?
> Do we have cases of claims that are will not have a corresponding annotation?
> What will be the UX of annotation that is mandatory and is not in the host's annotation? How does it reflect in your design?
> What will be the UX of annotation that has mapping but it's mapped value is not found? How does it reflect in your design?

1. Yes each host will need to have aud annotation is aud is mandatory
2. Yes that why we do the mapping. For example host will want to refer "ref" claim with the word "branch"
3. We fail the request. It will be reflected in the sequence diagram im adding now
4. If we have annotation that after mapping its not found we fail request. If we have mapping without relevant annotation we ignore it.

hilagross commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,178 @@
+| Name             | Value                   | Description                                      |
+| ---------------- | ----------------------- | ------------------------------------------------ |
+| mandatory_claims | branch, job, aud        | List of claims as they appear in host annotation |
+| claims_mapping   | {ref: branch, sub: job} | Mapping between host annotation to JWT claim     |
```
We decided without {}

tzheleznyak replied:
Done

hilagross commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,178 @@
+
+![image-20210711171238701](token-schema-class-diagram.png)
+
+### Flow Explanation
```
What will be the flow for mapping that is not mandatory?
What happen to the flow if the claim doesn't exists or not familiar (as it is host claim - not jwt claim)?

hilagross commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,178 @@
+## Backwards Compatibility
+[//]: # "How will the design of this solution impact backwards compatibility? Address how you are going to handle backwards compatibility, if necessary"
+
+Token schema is not mandatory therfore the JWT authentication will remain backwards compatible.
```
not sure we can use 'token schema' prase anymore, this is now 'mandatory fields and mapped claims' feature

hilagross commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,178 @@
+| Add status checks for token schema configuration | 2 SP       |
+| Add Cucumber tests for token schema              | 5 SP       |
+
+## Open Questions
```
I think we answered them today
please update this doc accordingly.

tzheleznyak replied:
Lets discuss it

hilagross commented on design/authenticators/authn_jwt/token_schema.md:
```diff
@@ -0,0 +1,178 @@
+
+| **Persona**        | **Name** |
+|--------------------|----------|
+| Team leader        |          |
+| Product owner      |          |
+| System architect   |          |
+| Security architect |          |
+| QA architect       |          |
```
by order:
@benoaviram 
@hilagross 
@tzheleznyak 
@sashaCher 
@eladkug

nessiLahav reviewed (approved):
LGTM
